### ISSUE TITLE
I recommend to increase batch sizes for powerful hardware (Dual Xeon)

### ISSUE THREAD

**Issue by Technologov** [glinscott/fishtest#978]
Hi,

I own several Dual Xeon Platinum 8176 servers, those are 56-core/112-thread machines.
The current batch size in Fishtest is only 200 games, according to the logs.

That means that every couple of minutes my servers re-download and re-compile Stockfish, a CPU effort that gets wasted, while it could be put to good use (...into more games being played).

I suggest that the batch size should depend on the hardware, with a small batch for Raspberry Pi, but a much bigger batches for Dual Xeon servers.

-Alexey "Technologov ", 01.08.2021

**Comment by Sopel97** [comment]
hmm, is there anything particularily hard preventing us from assigning, say, `min(cores*20, game_left_to_limit)` game batches?

**Comment by Technologov** [comment]
Let's try even bigger batches... 1 hour of computer time so (cores *60 ) for LTC games and (cores*60*6) for STC games

**Comment by Sopel97** [comment]
https://github.com/glinscott/fishtest/blob/55f4f9f66d618c6f627aec4222acb116cecb3b27/server/fishtest/rundb.py#L58

**Comment by Vizvezdenec** [comment]
I honestly don't remember why we even have such a small batches like 200.
With current fishtest mostly being run on 8+ core machines and also now some massive 50+ cores servers it would make a lot of sense to make batches bigger, like 1000-2000 games easily.

**Comment by Vizvezdenec** [comment]
@vondele

**Comment by ppigazzini** [comment]
@Technologov 
> That means that every couple of minutes my servers re-download and re-compile Stockfish, a CPU effort that gets wasted, while it could be put to good use (...into more games being played).

Wrong, the worker keeps the latest 50 binaries to optimize the download + build process, and the latest 10 nest to optimize the download, see
https://github.com/glinscott/fishtest/blob/9f4c421aee89e2eb92e3eddf270a994adec1958c/worker/games.py#L865-L873

https://github.com/glinscott/fishtest/blob/9f4c421aee89e2eb92e3eddf270a994adec1958c/worker/games.py#L937-L945

By the way the little VPS running the fishtest server scales very well, serving without problem 4803 machines at 29678 cores, see #553 where a SF developer whined "It's not that easy to beat speed of tests that complete if 20k cores are running - it's hard to produce any meaningful ideas at this tempo." :)

@Vizvezdenec 
> I honestly don't remember why we even have such a small batches like 200.

Here are the PR that changed the chunk_size and cached the binaries (#258), some reasons:

- FixedNumber and SPSA tests were stuck with workers at 1-2 cores
- IIRC when a SPRT test stopped the workers continued to play useless games wasting CPU

> With current fishtest mostly being run on 8+ core machines and also now some massive 50+ cores servers it would make a lot of sense to make batches bigger, like 1000-2000 games easily.

One months ago we had only 100 workers with low number of cores, and IIRC you cancelled some tests from the queue. I'm cheering to have a machines table full, though :)

Anyway these PRs introduced the games batching to lower the connections load to the server and use the chunk and batch size (#627, #628), change the chunk_size could break several things.

**Comment by Vizvezdenec** [comment]
We had 100 workers with like 1000 cores minimum, so average core count was 10.
Currently we have like 2k cores running from which : 
11 are single core workers 
5 are 2 core workers.
Everything else is 3+

**Comment by ppigazzini** [comment]
@Vizvezdenec feel free to open your first fishtest PR :)

**Comment by Vizvezdenec** [comment]
I want to hear feedback from Joost first, after all this stuff usually is approved by general maintainer of the project :)

**Comment by ppigazzini** [comment]
> hmm, is there anything particularily hard preventing us from assigning, say, `min(cores*20, game_left_to_limit)` game batches?

Only the lack of fishtest code contributors :)

At example now:
- the task list is created after creating a new test (IIRC)
- if a worker leaves, the unfinished task is assigned to the first free worker. So a task with a big number of games, say `min(cores*20, game_left_to_limit)`, could be assigned at 1 core worker: a problem for SPSA and FixedGames tests.

fishtest can (and should) be refactored and improved, simply some changes require a lot of work to introduce a new logic and to show correctly the tests result with either new or old logics.

**Comment by vondele** [comment]
I think some of the code depends on the batch size, so one would need to do such a change carefully, I don't think it is trivial to do on a per worker basis. As @ppigazzini pointed out, the overhead of the small batch size is quite minimal, as there is caching of the compiles.

**Comment by Vizvezdenec** [comment]
I think that currently with 2000+ cores we actually have tests being run pretty fast. 
In this case caching compiles probably doesn't help whatsoever because you are really often assigned to new jobs?

**Comment by NightlyKing** [comment]
I observe this on my machine as well. STC is scaled down to 4 seconds and it finishes the batch really quickly. Then it jumps to a new test that it hasn't seen yet and then again to a new one.

**Comment by jhellis3** [comment]
4 seconds, oh my.

**Comment by ppigazzini** [comment]
> I want to hear feedback from Joost first, after all this stuff usually is approved by general maintainer of the project :)

IMO @vondele was neutral, so you have green light to open the PR to change these couple of values :) (keep fixed "num_games")
https://github.com/glinscott/fishtest/blob/82ce909ef592cb0ff91652379b9b586eec400196/server/fishtest/rundb.py#L49
https://github.com/glinscott/fishtest/blob/82ce909ef592cb0ff91652379b9b586eec400196/server/fishtest/views.py#L666-L669

Requirements:
- SPRT batch size: (the num_games should be divisible by 8)
https://github.com/glinscott/fishtest/blob/82ce909ef592cb0ff91652379b9b586eec400196/server/fishtest/views.py#L651-L665
https://github.com/glinscott/fishtest/blob/82ce909ef592cb0ff91652379b9b586eec400196/server/fishtest/stats/stat_util.py#L166-L191
https://github.com/glinscott/fishtest/blob/82ce909ef592cb0ff91652379b9b586eec400196/worker/games.py#L1026-L1028
- SPSA batch size:
https://github.com/glinscott/fishtest/blob/82ce909ef592cb0ff91652379b9b586eec400196/worker/games.py#L1017-L1021

- we have a dozen of 1 core worker, so in the worst scenario we could have a dozen of SPSA/FixedGames slowly finishing the last task at 1.5 x LTC

I will test your PR on DEV with some of my workers.

**Comment by vondele** [comment]
what about excluding workers with less than e.g. 4 cores from FixedNumGames tests ?

**Comment by ppigazzini** [comment]
> what about excluding workers with less than e.g. 4 cores from FixedNumGames tests ?

Of course, that could be the second @Vizvezdenec PR, after the hard first one it should be a nice downhill :)

**Comment by Vizvezdenec** [comment]
basically I change 
self.chunk_size to 1000
and 
data["num_games"] = 400 * request.rundb.chunk_size 
?

**Comment by ppigazzini** [comment]
Yes :)

**Comment by Vizvezdenec** [comment]
okay I will try to make it tomorrow, would be faster if you remind me of how to change code for PR w/o having it to be compiled on my machine :D

**Comment by ppigazzini** [comment]
```
git clone https://github.com/Vizvezdenec/fishtest.git
cd fishtest
git remote add upstream https://github.com/glinscott/fishtest.git
git config user.email "Vizvezdenec@gmail.com"
git config user.name "Vizvezdenec"
git fetch -p origin
git fetch -p upstream
git reset --hard upstream/master
git push origin master
git checkout -b set_task_to_1000_games
sed -i 's/self.chunk_size = 200/self.chunk_size = 1000/' ./server/fishtest/rundb.py
sed -i 's/data\["num_games"\] = 2000/data\["num_games"\] = 400/' ./server/fishtest/views.py
git add .
git commit -m "Set task size to 1000 games" -m "bla
bla bla
bla bla bla
"
git push origin set_task_to_1000_games
```
then open the PR on GitHub

**Comment by ppigazzini** [comment]
Solved with dynamic task #1021 thanks to @vdbergh :)
